### PR TITLE
Libcaliptra piecewise FW load

### PR DESCRIPTION
**Pull request by nquarton**

### CODE REVIEW

**Review by jlmahowa-amd — DISMISSED** [review]

File: libcaliptra/examples/generic/test.c [hunk 1]
```diff
@@ -668,6 +668,91 @@ int rom_test_devid_csr(const test_info* info)
     return failure;
 }
 
+// Issue FW load commands repeatedly
+// Coverage for piecewise FW load and runtime FW updates
+int upload_fw_piecewise(const test_info* info)
+{
+    int failure = 0;
+    int status = boot_to_ready_for_fw(info, false);
```
Comment (fdamato) [inline libcaliptra/examples/generic/test.c:676]: shall we sanity_check info ?

Comment (nquarton) [reply libcaliptra/examples/generic/test.c:676]: If anything I would say to check it at run_tests() instead of each individual one. But I think either way the test is going to fail if someone breaks something.

File: libcaliptra/examples/generic/test.c [hunk 2]
```diff
@@ -668,6 +668,96 @@ int rom_test_devid_csr(const test_info* info)
...
+
+    if (status){
+        printf("Failed to boot to ready for FW: 0x%x\n", status);
+        failure = 1;
+    }
+
+    // Some "random" size to split up the FW load into chunks
+    // These represent the first two chunks, the third chunk is the remainder of the image
+    // Sizes of 0 are ignored (meaning one fewer chunk is sent)
+    uint32_t chunk_sizes[][2] =    {
```
Comment (fdamato) [inline libcaliptra/examples/generic/test.c:686]: Ideally the chunk should be randomly generated, and make sure they don`t add up for more space than the total image size...but I understand this is non production code...so it is not a sticky point

File: libcaliptra/examples/generic/test.c [hunk 3]
```diff
@@ -668,6 +668,96 @@ int rom_test_devid_csr(const test_info* info)
...
+        {
+            printf("FW Load %d End Failed: 0x%x\n", i, status);
+            dump_caliptra_error_codes();
+            failure = 1;
+        } else {
+            printf("FW Load %d End: OK\n", i);
+        }
+    }
+
+    return failure;
```
Comment (fdamato) [inline libcaliptra/examples/generic/test.c:759]: np: why carrying another flag here ? can`t we just return status ?

**Review by fdamato — APPROVED** [review]

File: libcaliptra/examples/generic/test.c (hunk 3 above)
Comment (nquarton) [reply libcaliptra/examples/generic/test.c:759]: These are written so the test will continue after a single failure but still report the result. So failure is sticky whereas status gets overwritten on each call.

File: libcaliptra/examples/generic/test.c (hunk 2 above)
Comment (nquarton) [reply libcaliptra/examples/generic/test.c:686]: I think the methodology I have seen in Caliptra is for deterministic testing over randomization which can cause flakeyness. I do agree that would be better coverage overall tough. (So long as the corner cases are for the code being tested and not corner cases in the test itself.)

**Review by fdamato — APPROVED** [review]
